R/match.R

```diff
@@ -108,9 +108,43 @@
 #'     `"last"` if you just need to detect if there is at least one match.
 #'   - `"first"` returns the first match detected in `haystack`.
 #'   - `"last"` returns the last match detected in `haystack`.
-#'   - `"warning"` throws a warning if multiple matches are detected, but
-#'     otherwise falls back to `"all"`.
-#'   - `"error"` throws an error if multiple matches are detected.
+#'
+#' @param relationship Handling of the expected relationship between
+#'   `needles` and `haystack`. If the expectations chosen from the list below
+#'   are invalidated, an error is thrown.
+#'
+#'   - `"none"` doesn't perform any relationship checks.
+#'
+#'   - `"one_to_one"` expects:
+#'     - Each value in `needles` matches at most 1 value in `haystack`.
+#'     - Each value in `haystack` matches at most 1 value in `needles`.
+#'
+#'   - `"one_to_many"` expects:
+#'     - Each value in `needles` matches any number of values in `haystack`.
+#'     - Each value in `haystack` matches at most 1 value in `needles`.
+#'
+#'   - `"many_to_one"` expects:
+#'     - Each value in `needles` matches at most 1 value in `haystack`.
+#'     - Each value in `haystack` matches any number of values in `needles`.
+#'
+#'   - `"many_to_many"` expects:
+#'     - Each value in `needles` matches any number of values in `haystack`.
+#'     - Each value in `haystack` matches any number of values in `needles`.
+#'
+#'     This performs no checks, and is identical to `"none"`, but is provided to
+#'     allow you to be explicit about this relationship if you know it exists.
+#'
+#'   - `"warn_many_to_many"` doesn't assume there is any known relationship, but
```

Comment:
Should `warn` be a suffix instead of a prefix? I think that's what I'd have expected to make it clearer it's a variant.

Reply:
I couldn't really decide, because it's not like `"many_to_many"` errors and `"many_to_many_warn"` would just downgrade that error to a warning.

They do fairly different things:
- `"many_to_many"` is looking for _violations of_ the many-to-many relationship, i.e. many-to-many _is_ expected. But that relationship can't be violated so it ends up being an expressive no-op (it makes more sense for `"many_to_one"` and the other variants)
- `"warn_many_to_many"` is looking for _the existance of_ a many-to-many relationship, i.e. many-to-many _is not_ expected, and it warns if it finds one.

So they didn't really feel like variants to me

I'm not picky about this though, so if you still feel like the suffix is better then I'll change it, but I'll wait to hear what you think

Reply:
No strong opinion either way, we can go with your first intuition.